### ISSUE TITLE
dorado v0.5.0: models 0.4.3 and  combinations of modification models 5mC_5hmC and 5mCG_5hmCG  stall  randomly

### ISSUE THREAD

**Issue by janpb** [nanoporetech/dorado#538]
When basecalling modified bases with `dorado` v0.5.0, using the modification models `5mC_5hmC` and `5mCG_5hmCG` stall basecalling randomly, i.e. sometimes `dorado` finishes basecalling, but mostly it just stalls and is not progressing. 

I assume `5mC_5hmC` is a subset of `5mCG_5hmCG` and therefore would expect  `dorado` would either warn about my useless modification setting or finish basecalling. However, the current randomness of the basecalling outcome makes it unclear if it is an user error (mine) or `dorado`.

- Using `6mA` and `5mC_5hmC` or `6mA` and  `5mCG_5hmCG` works. 
- Interestingly, passing the same model twice, e.g. `6mA`, stalls as well. I do not know if this is somehow related to #514. 

- OS: Ubuntu 20.04.6 LTS | Release:        20.04
- GPU:  NVIDIA H100 PCIe (81559MiB RAM)

Dorado: 
-dorado:   0.5.0+0d932c0+cu118
-libtorch: 2.0.0-ont
-minimap2: 2.25-r1173

## Testing
- Random waiting times between runs
- Changing order of modification models in command
- Always same pod5 file as input: single pod5 from an _Arabidopsis thaliana_ sequencing run 
- Always model v4.3.0
- Always sup
- Always same GPU
- Not shown: single methylation models or always working combination of `6mA` and `5mCG_5hmCG` or `5mC_5hmC`
- I show here a selection of runs to demonstrate  the randomness of a successful finish or progress before stalling. 

### Results
- stall

```
$: dorado basecaller --recursive --device cuda:1 sup@v4.3.0,5mCG_5hmCG,5mC_5hmC FAX73743_pass_ff688765_40d21d7f_14.pod5 > out.bam
[2023-12-18 21:53:22.349] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0 with httplib
[2023-12-18 21:53:24.710] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mCG_5hmCG@v1 with httplib
[2023-12-18 21:53:25.096] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mC_5hmC@v1 with httplib
[2023-12-18 21:53:25.472] [info] > Creating basecall pipeline
[2023-12-18 21:53:42.220] [info]  - set batch size for cuda:1 to 1792
███▏                         ] 14% [00m:34s<03m:31s]
^C
```

- stall

```
$: dorado basecaller --recursive --device cuda:1 sup@v4.3.0,5mCG_5hmCG,5mC_5hmC FAX73743_pass_ff688765_40d21d7f_11.pod5 > out.bam
[2023-12-18 21:58:08.523] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0 with httplib
[2023-12-18 21:58:11.008] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mCG_5hmCG@v1 with httplib
[2023-12-18 21:58:11.408] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mC_5hmC@v1 with httplib
[2023-12-18 21:58:11.799] [info] > Creating basecall pipeline
[2023-12-18 21:58:28.512] [info]  - set batch size for cuda:1 to 2048
[█▌                            ] 5% [00m:27s<08m:43s]
^C
```

- Finishing

```
$: dorado basecaller --recursive --device cuda:1 sup@v4.3.0,5mC_5hmC,5mCG_5hmCG FAX73743_pass_ff688765_40d21d7f_11.pod5 > out.bam
[2023-12-18 22:19:17.676] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0 with httplib
[2023-12-18 22:19:22.085] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mC_5hmC@v1 with httplib
[2023-12-18 22:19:22.854] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mCG_5hmCG@v1 with httplib
[2023-12-18 22:19:23.556] [info] > Creating basecall pipeline
[2023-12-18 22:19:40.267] [info]  - set batch size for cuda:1 to 2048
[2023-12-18 22:21:16.856] [info] > Simplex reads basecalled: 4000
[2023-12-18 22:21:16.856] [info] > Basecalled @ Samples/s: 8.714602e+06
[2023-12-18 22:21:16.902] [info] > Finished
```

- Finishing

```
$: dorado basecaller --recursive --device cuda:1 sup@v4.3.0,5mC_5hmC,5mCG_5hmCG  FAX73743_pass_ff688765_40d21d7f_11.pod5 > out.bam
[2023-12-18 22:24:06.773] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0 with httplib
[2023-12-18 22:24:09.292] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mC_5hmC@v1 with httplib
[2023-12-18 22:24:09.788] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mCG_5hmCG@v1 with httplib
[2023-12-18 22:24:10.328] [info] > Creating basecall pipeline
[2023-12-18 22:24:27.073] [info]  - set batch size for cuda:1 to 1856
[2023-12-18 22:26:03.189] [info] > Simplex reads basecalled: 4000
[2023-12-18 22:26:03.189] [info] > Basecalled @ Samples/s: 8.757494e+06
[2023-12-18 22:26:03.227] [info] > Finished
```

- stall

```
$: dorado basecaller --recursive --device cuda:1 sup@v4.3.0,5mCG_5hmCG,5mC_5hmC  FAX73743_pass_ff688765_40d21d7f_11.pod5 > out.bam
[2023-12-18 22:29:03.943] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0 with httplib
[2023-12-18 22:29:06.397] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mCG_5hmCG@v1 with httplib
[2023-12-18 22:29:06.795] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mC_5hmC@v1 with httplib
[2023-12-18 22:29:07.182] [info] > Creating basecall pipeline
[2023-12-18 22:29:24.152] [info]  - set batch size for cuda:1 to 1856
^C██████████████▎              ] 50% [01m:07s<01m:05s]
```

- Finished: changing order of methylation models 

```
$: dorado basecaller --recursive --device cuda:1 sup@v4.3.0,5mC_5hmC,5mCG_5hmCG  FAX73743_pass_ff688765_40d21d7f_11.pod5 > out.bam
[2023-12-18 22:37:34.057] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0 with httplib
[2023-12-18 22:37:36.413] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mC_5hmC@v1 with httplib
[2023-12-18 22:37:36.817] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mCG_5hmCG@v1 with httplib
[2023-12-18 22:37:37.196] [info] > Creating basecall pipeline
[2023-12-18 22:37:53.934] [info]  - set batch size for cuda:1 to 1920
[2023-12-18 22:39:30.413] [info] > Simplex reads basecalled: 4000
[2023-12-18 22:39:30.413] [info] > Basecalled @ Samples/s: 8.724539e+06
[2023-12-18 22:39:30.459] [info] > Finished
```

- A later rerun with this modification model order did stall:
```
$: dorado basecaller --recursive --device cuda:1 sup@v4.3.0,5mCG_5hmCG,5mC_5hmC FAX73743_pass_ff688765_40d21d7f_11.pod5 > out.bam
[2023-12-19 00:03:11.884] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0 with httplib
[2023-12-19 00:03:13.969] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mCG_5hmCG@v1 with httplib
[2023-12-19 00:03:14.313] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mC_5hmC@v1 with httplib
[2023-12-19 00:03:14.663] [info] > Creating basecall pipeline
[2023-12-19 00:03:31.372] [info]  - set batch size for cuda:1 to 1984
^C██▌                          ] 11% [00m:32s<03m:57s] 
```

- Stall

```
dorado basecaller --recursive --device cuda:1 sup@v4.3.0,5mCG_5hmCG,5mC_5hmC FAX73743_pass_ff688765_40d21d7f_11.pod5 > out.bam
[2023-12-18 22:40:01.025] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0 with httplib
[2023-12-18 22:40:03.343] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mCG_5hmCG@v1 with httplib
[2023-12-18 22:40:03.757] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mC_5hmC@v1 with httplib
[2023-12-18 22:40:04.134] [info] > Creating basecall pipeline
[2023-12-18 22:40:21.017] [info]  - set batch size for cuda:1 to 1728
^C████████▎                    ] 31% [00m:57s<02m:08s]
```

-stall:  Adding 6mA model 

```
dorado basecaller --recursive --device cuda:1 sup@v4.3.0,5mC_5hmC,5mCG_5hmCG,6mA FAX73743_pass_ff688765_40d21d7f_11.pod5 > out .bam
[2023-12-18 23:02:45.992] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0 with httplib
[2023-12-18 23:02:48.891] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mC_5hmC@v1 with httplib
[2023-12-18 23:02:49.453] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mCG_5hmCG@v1 with httplib
[2023-12-18 23:02:49.972] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_6mA@v1 with httplib
[2023-12-18 23:02:50.461] [info] > Creating basecall pipeline
[2023-12-18 23:03:07.368] [info]  - set batch size for cuda:1 to 1728
^C▌                            ] 5% [00m:30s<09m:34s]
```

- stall: Passing same model twice (6mA alone works, see below)
```
$: dorado basecaller --recursive --device cuda:1 sup@v4.3.0,6mA,6mA  FAX73743_pass_ff688765_40d21d7f_11.pod5 > out.bam
[2023-12-18 23:28:48.068] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0 with httplib
[2023-12-18 23:28:50.387] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_6mA@v1 with httplib
[2023-12-18 23:28:50.752] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_6mA@v1 with httplib
[2023-12-18 23:28:51.111] [info] > Creating basecall pipeline
[2023-12-18 23:29:07.794] [info]  - set batch size for cuda:1 to 1920
^C██▌                          ] 11% [00m:46s<05m:42s]
```

- Finishing: 6mA alone

```
$: dorado basecaller --recursive --device cuda:1 sup@v4.3.0,6mA FAX73743_pass_ff688765_40d21d7f_11.pod5 > out.bam
[2023-12-18 23:56:07.510] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0 with httplib
[2023-12-18 23:56:09.580] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_6mA@v1 with httplib
[2023-12-18 23:56:09.923] [info] > Creating basecall pipeline
[2023-12-18 23:56:26.542] [info]  - set batch size for cuda:1 to 2048
[2023-12-18 23:59:12.299] [info] > Simplex reads basecalled: 4000
[2023-12-18 23:59:12.299] [info] > Basecalled @ Samples/s: 5.077839e+06
[2023-12-18 23:59:12.351] [info] > Finished
```

**Comment by ymcki** [comment]
"I assume 5mC_5hmC is a subset of 5mCG_5hmCG". Isn't the reverse true instead?

**Comment by janpb** [comment]
@ymcki: You're right, I mixed this up.  Thanks for pointing it out.

**Comment by pustoshilov-d** [comment]
Hi @janpb 
Yes, that's the same that I got in https://github.com/nanoporetech/dorado/issues/514
If you set --verbose, you'll see `CuBLAS error 13` and `Aborted (core dumped) `
And yes, 5mCG_5hmCG is a subset of 5mC_5hmC https://github.com/nanoporetech/dorado/issues/519

So temp solution will be to use `5mC_5hmC,6mA`, if you have resources. Results will be same and even better

It would be great if developers add exeptions on this or add more info to Readme at least

**Comment by janpb** [comment]
@pustoshilov-d : I envy you as you are getting error messages at all.  I do not get any error message, nothing in stdout/stderr,  CuBLAS or in `dmesg`, with or without `--verbose`. Therefore I'm not sure it's the same issue (although expected to be at least  very closely related).

I would also expect that serious errors like  `CuBLAS error 13` and `Aborted (core dumped)` would be printed to `stderr/stdout` and/or logged in `dmesg` or similar, independently of using `--verbose`
```
$: dorado basecaller --verbose --recursive --device cuda:1 sup@v4.3.0,5mCG_5hmCG,5mC_5hmC FAX73743_pass_ff688765_40d21d7f_14.pod5 >out.bam 2> out.log

// content of out.log (no progress bar)
[2023-12-19 09:20:11.817] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0 with httplib
[2023-12-19 09:20:14.001] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mCG_5hmCG@v1 with httplib
[2023-12-19 09:20:14.350] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mC_5hmC@v1 with httplib
[2023-12-19 09:20:14.706] [info] > Creating basecall pipeline
[2023-12-19 09:20:15.503] [debug] cuda:1 memory available: 83.80GB
[2023-12-19 09:20:15.503] [debug] Auto batchsize cuda:1: memory limit 82.80GB
[2023-12-19 09:20:15.503] [debug] Auto batchsize cuda:1: testing up to 3392 in steps of 64
[2023-12-19 09:20:15.616] [debug] Auto batchsize cuda:1: 64, time per chunk 0.794896 ms
[2023-12-19 09:20:15.720] [debug] Auto batchsize cuda:1: 128, time per chunk 0.402443 ms
[2023-12-19 09:20:15.825] [debug] Auto batchsize cuda:1: 192, time per chunk 0.274482 ms
//snip
[2023-12-19 09:20:30.829] [debug] Auto batchsize cuda:1: 3328, time per chunk 0.079630 ms
[2023-12-19 09:20:31.377] [debug] Auto batchsize cuda:1: 3392, time per chunk 0.079092 ms
[2023-12-19 09:20:31.377] [debug] Device cuda:1 Model memory 32.75GB
[2023-12-19 09:20:31.377] [debug] Device cuda:1 Decode memory 13.54GB
[2023-12-19 09:20:31.427] [info]  - set batch size for cuda:1 to 1920
[2023-12-19 09:20:31.427] [debug] - adjusted chunk size to match model stride: 10000 -> 9996
[2023-12-19 09:20:31.429] [debug] - adjusted overlap to match model stride: 500 -> 498
[2023-12-19 09:20:33.367] [debug] Checking adapter/primer LSK109
[2023-12-19 09:20:33.367] [debug] Checking adapter/primer LSK110
[2023-12-19 09:20:33.367] [debug] Checking adapter/primer PCR_PSK_rev1
//snip
[2023-12-19 09:21:13.115] [debug] Checking adapter/primer cDNA_VNP
[2023-12-19 09:21:13.115] [debug] Checking adapter/primer cDNA_SSP
[2023-12-19 09:21:13.115] [debug] Checking adapter/primer PCS110_forward
[2023-12-19 09:21:13.115] [debug] Checking adapter/primer PCS110_reverse
//stuck 
```

**Comment by ymcki** [comment]
So we can get both 6mA and 5mC_5hmC in one run by specifying both models and adding one comma in between??

**Comment by pustoshilov-d** [comment]
@ymcki yes, look at https://github.com/nanoporetech/dorado#automatic-model-selection-complex

**Comment by janpb** [comment]
I went digging in `dmesg` and  found only these two messages in which **may be** related to the issue based on the time (and dorado). the PCI belongs to the GPU used for testing.   I did not log the pids of my runs, tough. So this has to be taken with a grain of salt.  

```
[Mon Dec 18 21:13:46 2023] NVRM: Xid (PCI:0000:81:00): 31, pid=75971, name=dorado, Ch 00000008, intr 00000000. MMU Fault: ENGINE GRAPHICS GPCCLIENT_T1_7 faulted @ 0x7fd4_ce540000. Fault is of type FAULT_PDE ACCESS_TYPE_VIRT_WRITE
[Mon Dec 18 21:52:08 2023] NVRM: Xid (PCI:0000:81:00): 31, pid=77015, name=dorado, Ch 00000008, intr 00000000. MMU Fault: ENGINE GRAPHICS GPCCLIENT_T1_12 faulted @ 0x7f0e_8be60000. Fault is of type FAULT_PDE ACCESS_TYPE_VIRT_WRITE
```
Xid 31 can indicate an illegal memory access: https://forums.developer.nvidia.com/t/fault-is-of-type-fault-pde-access-type-virt-read-kernel-problem-or-driver-issue/180325/2

**Comment by tijyojwad** [comment]
Hi @janpb - running 2 models that cover the same context isn't supported correctly, and I agree dorado needs to report/handle this correctly. We're looking into adding this error handling now. Apologies for the problem you're running into!

**Comment by tijyojwad** [comment]
Hi @janpb - we just released v0.5.1 (https://github.com/nanoporetech/dorado#installation) which now prevents users from running models which predict mods on overlapping canonical bases.
